core/src/prediction/collector/OptimizedPredictionCollector.cpp

```diff
@@ -30,15 +32,62 @@ std::vector<Prediction> OptimizedPredictionCollector::collect_predictions(const
                                                                           const std::vector<std::vector<size_t>>& leaf_nodes_by_tree,
                                                                           const std::vector<std::vector<bool>>& valid_trees_by_sample,
                                                                           bool estimate_variance,
-                                                                          bool estimate_error) const {
-  size_t num_trees = forest.get_trees().size();
+                                                                          bool estimate_error,
+                                                                          uint num_threads) const {
   size_t num_samples = data.get_num_rows();
+  std::vector<uint> thread_ranges;
+  split_sequence(thread_ranges, 0, num_samples - 1, num_threads);
+
+  std::vector<std::future<std::vector<Prediction>>> futures;
+  futures.reserve(thread_ranges.size());
+
+  std::vector<Prediction> predictions;
+  predictions.reserve(num_samples);
+
+  for (uint i = 0; i < thread_ranges.size() - 1; ++i) {
+    size_t start_index = thread_ranges[i];
+    size_t num_samples_batch = thread_ranges[i + 1] - start_index;
+
+    futures.push_back(std::async(std::launch::async,
+                                 &OptimizedPredictionCollector::collect_predictions_batch,
+                                 this,
+                                 std::ref(forest),
+                                 std::ref(train_data),
+                                 std::ref(data),
+                                 std::ref(leaf_nodes_by_tree),
+                                 std::ref(valid_trees_by_sample),
+                                 estimate_variance,
+                                 estimate_error,
+                                 start_index,
+                                 num_samples_batch));
+  }
+
+  for (auto& future : futures) {
+    std::vector<Prediction> thread_predictions = future.get();
+    predictions.insert(predictions.end(),
+                       std::make_move_iterator(thread_predictions.begin()),
+                       std::make_move_iterator(thread_predictions.end()));
```

Comment:
A note though: the speed is the same without `std::make_move_iterator`, suggesting `Prediction`'s (autogenerated??) move operator is not called?

Reply:
There are some unnecessary copies going on here. I think collect_predictions_batch should return a pointer to the vector of predictions.

Reply:
3d0c763 has the thread-wise function return a pointer to the possibly "large" object returned. It made no speed difference, but feels more "right"?

This is the diagram of what is happening inside this loop, are there more efficient ways of getting the guys on the right over to the Prediction object on the left?

```
             +-----------------+                                                 +---------------+
             |                 |                                                 |               |
             |                 |                                                 |  Prediction   |
             |                 |                                                 |               |
             |    Prediction   |                                                 |               |
             |    (empty)      |                                                 +---------------+
             |                 |                                 ptr1 +------>   |               |
             |                 |                                                 |               |
             |                 |                                                 |  Prediction   |
             +-----------------+                                                 |               |
             |                 |                <-----------+                    |               |
             |                 |                                                 +---------------+
             |                 |                                                 |               |
             |      ...        |                                                 |               |
             |                 |                                                 |  Prediction   |
             |                 |                                                 |               |
             |                 |                                                 |               |
Predictions  +-----------------+                                                 +---------------+
             |                 |
             |                 |
             |                 |                                                 +---------------+
             |                 |                                                 |               |
             |                 |                                                 |  Prediction   |
             |                 |                                                 |               |
             |                 |                                                 |               |
             |                 |                                                 +---------------+
             +-----------------+                                 ptr2 +------>   |               |
             |                 |                                                 |               |
             |                 |                                                 |  Prediction   |
             |                 |                <-----------+                    |               |
             |                 |                                                 |               |
             |                 |                                                 +---------------+
             |                 |                                                 |               |
             |                 |                                                 |               |
             |                 |                                                 |  Prediction   |
             |                 |                                                 |               |
             |                 |                                                 |               |
             |                 |                                                 +---------------+
             |                 |
             +-----------------+

```

Reply:
I'm not sure I totally follow -- maybe we can keep this as-is for now (since it's consistent with the strategy we take elsewhere) and we can plan to discuss the use of move/ copy in person.